cmd/buildctl/build.go

```diff
@@ -315,9 +317,61 @@ func buildAction(clicontext *cli.Context) error {
 }
 
 func writeMetadataFile(filename string, exporterResponse map[string]string) error {
-	b, err := json.Marshal(exporterResponse)
+	var err error
+	out := make(map[string]interface{})
+	for k, v := range exporterResponse {
+		dt, err := base64.StdEncoding.DecodeString(v)
+		if err != nil {
+			out[k] = v
+			continue
+		}
+		raw := make(map[string]json.RawMessage)
+		if err = json.Unmarshal(dt, &raw); err != nil {
+			out[k] = v
+			continue
+		}
+		if len(raw) == 0 {
+			continue
+		}
+		out[k] = raw
+	}
+	b, err := json.Marshal(out)
 	if err != nil {
 		return err
 	}
-	return continuity.AtomicWriteFile(filename, b, 0666)
+	return continuity.AtomicWriteFile(filename, stripJSONNullProperties(b), 0666)
+}
+
+func stripJSONNullProperties(src []byte) []byte {
```

Comment:
That's not what I meant. Just `json.Unmarshal` into a `map[string]interface{}`. If it decodes without an error and `len(map) > 0` then add the bytes to out as `RawMessage`. Otherwise, we show the value as the original string.

Reply:
Sure. For the strip json null props func I was thinking it could be useful for cases like `{"foo":null,"bar":"baz"}`.

Reply:
I'm not sure what case it is? If the daemon already sends JSON then we should not transform it.

Reply:
Was just preventive but agree we should not change it.